[SERV-1106] Add PTLS tool to pt-tools

angelahuqing opened the pull request:
- This PR includes much of the structure of how other command line tools will be written for PT-tools
- PT-tools has both an api side as well as a cli side 
- the api is contained in pkg/pairtree
- the cli is called from main.go which accesses the cli in cmd/ptls
- Testing for both are in the same folders (as is normal in go structure). Testing uses testifys assert as well as afero to create temporary directories
- test-dir contains the test pairtree directory that most testing is done on. This test-dir is copied in various tests
- pkg/errror-msgs contains error-msgs that are used through out the program
- pkg/test-utls has various testing functoins that are used in different test spread through different packages 
- There will likely be things I change as we progress further in the project. I'm debating how best to deal with the Logger , but for now, having it in the ptls package works and is simple

angelahuqing committed:
[SERV-1106] Create ptls and testing files

angelahuqing committed:
Merge remote-tracking branch 'origin/main' into SERV-1106

angelahuqing committed:
[SERV-1106] Update structure of PTLS and include testing

ksclarke commented on pkg/pairtree/pairtree.go:
```diff
@@ -0,0 +1,243 @@
+
+	ptRoot = filepath.Join(ptRoot, rootDir)
+
+	pairPath := caltech_pairtree.Encode(id)
```
I'm guessing `func EncodeChar(id string) string {}` is necessary because `caltech_pairtree.Encode()` includes the splitting into pairs, and there isn't a separate encoding function in that library independent from that function?

angelahuqing replied:
The `caltech_pairtree.Encode()` only splits into pairs and doesn't deal with or do anything with special characters which is a part of the pairtree spec. The encode and decode functions from caltech are incredibly basic and only split things into pairs and nothing else. That's why there are multiple functions that build the pairtree path outside of caltech. Caltech isn't actually encoding anything.

angelahuqing replied:
We are waiting for Caltech to get back to us on making their EncodeChar() function public.

ksclarke reviewed (commented):
We need build instructions in the README. I know we didn't use ACT for go-festerize because we hadn't adopted it at that point, but is there a reason not to use it here, now?

angelahuqing commented on pkg/error-msgs/error_msgs.go:
```diff
@@ -0,0 +1,13 @@
+import "errors"
+
+var (
+	Err1 = errors.New("pairtree_prefix file exists, but is empty and must be populated")
```
I'm neutral either way. In our java projects we have them numbered and not named, so this is to keep consistent with that.

angelahuqing replied:
We decided to keep the errors numbered for ease of use.

angelahuqing commented on README.md:
```diff
@@ -4,21 +4,25 @@ This is a Go command line program that will allow listing and interacting with a
 or when the ENV PAIRTREE_ROOT is not set 
 
-    ptls [PT_ROOT] "[ID]"
+    ptls --Root [PT_ROOT] "[ID]"
```
`-R` is the short form. That's a mistake in the readme.  `-p` and `--pairtree_root` aren't used in the CLI?

angelahuqing committed:
[SERV-1106] Fix checkstyle issues and update program with small cahnges

angelahuqing committed:
[SERV-1106] Update README with build instructions

angelahuqing committed:
[SERV-1106] Update README

angelahuqing committed:
[SERV-1106] Update build.yml to run tests

angelahuqing committed:
[SERV-1106] Update build.yml

angelahuqing committed:
[SERV-1106] Update tests and test pairtree

There was an issue with the .hidden folder in b5488 because there was no file in it

angelahuqing committed:
Update fielpaths to work with different OS

Replace EncodeChar with CharEncode from CalTech

angelahuqing committed:
[SERV-1106] Update GA and tests with new hidden file

angelahuqing commented:
> We need build instructions in the README. I know we didn't use ACT for go-festerize because we hadn't adopted it at that point, but is there a reason not to use it here, now?

We're going to use Make since it seems to be the most common tool in GO for building. Another ticket has been created for this.

DRickard commented on pkg/.DS_Store:
Are the DS_Store files necessary for the repo, or could they be added to .gitignore?

ksclarke commented on README.md:
```diff
@@ -2,23 +2,37 @@
 
+## Installation
+
+First, ensure that you have Go Version 1.22 on you system. Clone this repository
```
I wonder if there is a way to supply this version number automatically? Javascript that pulls down the version from DockerHub's latest container? Or maybe just saying "the latest" version? I've had READMEs before with these sorts of version numbers and they become a pain to manually keep up with (e.g., I think Go is on 1.23 now?) I don't think this is a blocker for this PR, but something we might want to discuss/think about.

angelahuqing committed:
[SERV-1106] Remove .DS_Store and add to .gitignore
